core/vm/ipgraph_with_kind.go

```diff
@@ -0,0 +1,365 @@
+
+	// Initialize result to accumulate the final royalty percentage
+	result := big.NewInt(0)
+
+	// Clear any previous state related to ancestors
+	ancestorIps := make(map[common.Address]struct{})
+	childMap := make(map[common.Address]common.Address)
+
+	// Stack for DFS traversal
+	var stack []common.Address
+	stack = append(stack, ipId)
+
+	// Perform DFS to find the path from ipId to ancestorIpId
+	for len(stack) > 0 {
+		currentIpId := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+
+		// If we've reached the ancestor, stop the search
+		if currentIpId == ancestorIpId {
+			break
+		}
+
+		// Get the number of parent IPs for the current node
+		currentLengthHash := evm.StateDB.GetState(ipGraphAddress, common.BytesToHash(currentIpId.Bytes()))
+		currentLength := currentLengthHash.Big()
+
+		// If there are no parent IPs, continue to the next node
+		if currentLength.Uint64() == 0 {
+			continue
+		}
+
+		// Traverse all parent IPs
+		for i := uint64(0); i < currentLength.Uint64(); i++ {
+			slot := crypto.Keccak256Hash(currentIpId.Bytes()).Big()
+			slot.Add(slot, new(big.Int).SetUint64(i))
+			storedParent := evm.StateDB.GetState(ipGraphAddress, common.BigToHash(slot))
+			parentIpId := common.BytesToAddress(storedParent.Bytes())
+
+			// Map the parent to the child to reconstruct the path later
+			childMap[parentIpId] = currentIpId
+
+			// If this parent hasn't been visited, add it to the stack
+			if _, found := ancestorIps[parentIpId]; !found {
+				ancestorIps[parentIpId] = struct{}{}
+				stack = append(stack, parentIpId)
+			}
+		}
+	}
+
+	// Reconstruct the path from ipId to ancestorIpId and calculate royalty
+	currentIpId := ancestorIpId
+	var lrpQueue []common.Address
+
+	// Traverse the childMap to find the full path from ipId to ancestorIpId
+	for currentIpId != ipId {
+		lrpQueue = append(lrpQueue, currentIpId)
+		currentIpId = childMap[currentIpId]
+	}
+
+	// If a path is found, calculate the royalty along the path
+	currentIpId = ipId
+	if len(lrpQueue) > 0 {
+		// Pop the first parent from the queue and calculate initial royalty
+		parentIpId := lrpQueue[len(lrpQueue)-1]
+		lrpQueue = lrpQueue[:len(lrpQueue)-1]
+
+		royaltySlot := crypto.Keccak256Hash(currentIpId.Bytes(), parentIpId.Bytes(), royaltyPolicyKindLRP.Bytes()).Big()
+		initialRoyalty := evm.StateDB.GetState(ipGraphAddress, common.BigToHash(royaltySlot)).Big()
+		result.Set(initialRoyalty)
+		currentIpId = parentIpId
+
+		// Multiply royalties along the remaining path
+		for len(lrpQueue) > 0 {
+			parentIpId = lrpQueue[len(lrpQueue)-1]
+			lrpQueue = lrpQueue[:len(lrpQueue)-1]
+
+			royaltySlot = crypto.Keccak256Hash(currentIpId.Bytes(), parentIpId.Bytes(), royaltyPolicyKindLRP.Bytes()).Big()
+			nextRoyalty := evm.StateDB.GetState(ipGraphAddress, common.BigToHash(royaltySlot)).Big()
+
+			// Multiply the result by the next royalty and divide by 10000 (assuming basis points)
```

Comment:
How can we make sure the royalties are basis points? from the contracts? any chance we can check here?

Reply:
It is constant and we currently have e2e tests to verify it from contract level.